File: Commons/Data/src/main/java/com/greentree/commons/data/FileWatcher.java
package com.greentree.commons.data;

import com.greentree.commons.action.CloseRunnable;
import com.greentree.commons.action.ListenerCloser;
import com.greentree.commons.action.observable.ObjectObservable;
import com.greentree.commons.action.observer.object.EventAction;
import com.greentree.commons.data.resource.ResourceAction;
import com.greentree.commons.data.resource.ResourceActionImpl;
import com.greentree.commons.util.collection.FunctionAutoGenerateMap;
import com.greentree.commons.util.exception.WrappedException;

import java.io.File;
import java.io.IOException;
import java.lang.ref.Cleaner;
import java.nio.file.FileSystems;
import java.nio.file.Path;
import java.nio.file.WatchKey;
import java.nio.file.WatchService;
import java.util.Map;

import static java.nio.file.StandardWatchEventKinds.*;

public class FileWatcher {

    private final static Map<Path, FolderResourceAction> folders = FunctionAutoGenerateMap
            .build(FolderResourceAction::new);

    private FileWatcher() {
    }

    public static ResourceAction getFileAction(File file) {
        return getFileAction(file.toPath());
    }

    public static ResourceAction getFileAction(Path file) {
        return new FileResourceAction(folders.get(file.getParent()), file);
    }

    public static FolderResourceAction getFolderAction(File dir) {
        assert dir.exists() && dir.isDirectory() || !dir.exists() : dir + " is not file";
        return getFolderAction(dir.toPath());
    }

    public static FolderResourceAction getFolderAction(Path dir) {
        return folders.get(dir);
    }

    public static void pollEvents() {
        for (var a : folders.values())
            a.pollEvents();
    }

    public static void takeEvents() {
        for (var a : folders.values())
            a.takeEvents();
    }

    private static final class FileResourceAction extends ResourceActionImpl {

        private static final long serialVersionUID = 1L;

        private final ListenerCloser lc1, lc2, lc3;

        public FileResourceAction(FolderResourceAction action, Path file) {
            lc1 = action.getOnCreateAbsolute().addListener(f -> {
                if (f.equals(file))
                    eventCreate();
            });
            lc2 = action.getOnModifyAbsolute().addListener(f -> {
                if (f.equals(file))
                    eventModify();
            });
            lc3 = action.getOnDeleteAbsolute().addListener(f -> {
                if (f.equals(file))
                    eventDelete();
            });
            var cleaner = Cleaner.create();
            cleaner.register(this, new CloseRunnable(lc1));
            cleaner.register(this, new CloseRunnable(lc2));
            cleaner.register(this, new CloseRunnable(lc3));
        }

    }

    public static final class FolderResourceAction {

        private transient final EventAction<Path> onCreate = new EventAction<>();
        private transient final EventAction<Path> onModify = new EventAction<>();
        private transient final EventAction<Path> onDelete = new EventAction<>();
        private transient final EventAction<Path> onCreateAbsolute = new EventAction<>();
        private transient final EventAction<Path> onModifyAbsolute = new EventAction<>();
        private transient final EventAction<Path> onDeleteAbsolute = new EventAction<>();
        private final WatchService watchService;
        private final Path dir;

        public FolderResourceAction(Path dir) throws IOException {
            try {
                watchService = FileSystems.getDefault().newWatchService();
            } catch (IOException e) {
                throw new WrappedException(e);
            }
            dir.register(watchService, ENTRY_CREATE, ENTRY_MODIFY, ENTRY_DELETE);
            this.dir = dir;
        }

        public Path getDir() {
            return dir;
        }

        public ObjectObservable<Path> getOnCreate() {
            return onCreate;
        }

        public ObjectObservable<Path> getOnCreateAbsolute() {
            return onCreateAbsolute;
        }

        public ObjectObservable<Path> getOnDelete() {
            return onDelete;
        }

        public ObjectObservable<Path> getOnDeleteAbsolute() {
            return onDeleteAbsolute;
        }

        public ObjectObservable<Path> getOnModify() {
            return onModify;
        }

        public ObjectObservable<Path> getOnModifyAbsolute() {
            return onModifyAbsolute;
        }

        public boolean pollEvents() {
            try {
                final var key = watchService.poll();
                if (key != null)
                    return solveKey(key);
                return true;
            } catch (Exception e) {
                throw new WrappedException(e);
            }
        }

        private boolean solveKey(WatchKey key) {
            for (var event : key.pollEvents())
                try {
                    final var path = (Path) event.context();
                    final var absolute = dir.resolve(path);
                    if (event.kind() == ENTRY_CREATE) {
                        onCreate.event(path);
                        onCreateAbsolute.event(absolute);
                    }
                    if (event.kind() == ENTRY_MODIFY) {
                        onModify.event(path);
                        onModifyAbsolute.event(absolute);
                    }
                    if (event.kind() == ENTRY_DELETE) {
                        onDelete.event(path);
                        onDeleteAbsolute.event(absolute);
                    }
                } catch (Exception e) {
                    throw new WrappedException(e);
                }
            return key.reset();
        }

        public boolean takeEvents() {
            try {
                final var key = watchService.take();
                solveKey(key);
                return key != null;
            } catch (Exception e) {
                throw new WrappedException(e);
            }
        }

        @Override
        public String toString() {
            return "FolderResourceAction [" + dir + "]";
        }

    }

}

File: Commons/Data/src/main/java/com/greentree/commons/data/FileWatcher.kt
package com.greentree.commons.data

import com.greentree.commons.action.ListenerCloser
import com.greentree.commons.action.container.MultiContainer
import com.greentree.commons.action.observable.RunObservable
import java.io.File
import java.nio.file.FileSystems
import java.nio.file.Path
import java.nio.file.StandardWatchEventKinds.*
import java.nio.file.WatchEvent
import java.nio.file.WatchKey
import java.nio.file.WatchService
import java.util.concurrent.TimeUnit
import kotlin.io.path.absolute
import kotlin.io.path.exists

interface FileKeyWatcher {

	fun take()

	fun poll(): Boolean
	fun poll(timeout: Long, unit: TimeUnit): Boolean
}

data object FileWatcher : FileKeyWatcher {

	private val watchers = MultiContainer<FileKeyWatcher>()

	private fun onFile(file: File, vararg events: WatchEvent.Kind<*>) = onFile(file.toPath(), *events)

	private fun onFile(file: Path, vararg events: WatchEvent.Kind<*>): RunObservable {
		require(!file.exists() || file.toFile().isFile)
		return OnKindFileWatcher(
			file.absolute(),
			arrayOf(*events)
		)
	}

	fun onFileModify(file: File) = onFile(file, ENTRY_MODIFY)
	fun onFileModify(file: Path) = onFile(file, ENTRY_MODIFY)

	fun onFileCreate(file: File) = onFile(file, ENTRY_CREATE)
	fun onFileCreate(file: Path) = onFile(file, ENTRY_CREATE)

	fun onFileDelete(file: File) = onFile(file, ENTRY_DELETE)
	fun onFileDelete(file: Path) = onFile(file, ENTRY_DELETE)

	class OnKindFileWatcher(
		val file: Path,
		val events: Array<WatchEvent.Kind<*>>,
	) : RunObservable {

		val directory: Path
			get() = file.parent

		init {
			require(file.isAbsolute)
		}

		override fun addListener(listener: Runnable): ListenerCloser {
			val watchService: WatchService = FileSystems.getDefault().newWatchService()
			val thisKey = directory.register(watchService, events)
			val lc = watchers.add(object : FileKeyWatcher {
				override fun take() {
					val key = watchService.take()
					resolveKey(key)
				}

				override fun poll(): Boolean {
					val key = watchService.poll()
					if(key != null) {
						resolveKey(key)
						return true
					}
					return false
				}

				override fun poll(timeout: Long, unit: TimeUnit): Boolean {
					val key = watchService.poll(timeout, unit)
					if(key != null) {
						resolveKey(key)
						return true
					}
					return false
				}

				fun resolveKey(key: WatchKey) {
					for(event in key.pollEvents()) {
						if(event.kind() !in events)
							continue
						val path = directory.resolve(event.context() as Path)
						if(file == path)
							listener.run()
					}
					key.reset()
				}
			})
			return ListenerCloser {
				lc.close()
				thisKey.cancel()
				watchService.close()
			}
		}
	}

	override fun take() {
		while(!poll()) {
		}
	}

	override fun poll(): Boolean {
		var result = false
		for(watcher in watchers) {
			result = watcher.poll() || result
		}
		return result
	}

	override fun poll(timeout: Long, unit: TimeUnit): Boolean {
		val time = System.currentTimeMillis() + unit.toMillis(timeout)
		var result = false
		while(!result || time > System.currentTimeMillis()) {
			for(watcher in watchers) {
				result = watcher.poll(timeout, unit) || result
			}
		}
		return result
	}
}
